webapp/app.py

```diff
@@ -114,3 +121,36 @@ def document(path):
         updated=document["updated"],
         nav_html=nav_html,
     )
+
+
+@app.route("/search")
+def search():
+    """
+    Get search results from Google Custom Search
+    """
+    search_api_key = current_app.config["SEARCH_API_KEY"]
+    search_api_url = current_app.config["SEARCH_API_URL"]
+    search_custom_id = current_app.config["SEARCH_CUSTOM_ID"]
+
+    query = request.args.get("q")
+    site = request.args.get("siteSearch") or request.args.get("domain")
+    num = int(request.args.get("num", "10"))
+    start = int(request.args.get("start", "1"))
+
+    context = {"query": query, "start": start, "siteSearch": site, "num": num}
+
+    if query:
+        try:
+            context["results"] = get_search_results(
+                search_api_key,
+                search_api_url,
+                search_custom_id,
+                query,
+                start,
+                num,
+                site,
+            )
+        except Exception:
```

Comment:
This confused me for a long time - why this wasn't showing me an error in the logs, or in the browser.

Ideally we should be as specific as we can with catching exceptions. If we want to catch a specific error to do something specific that's fine, but otherwise we should let the standard exception handling logic take over, so we get proper tracebacks etc.

And in this case, if all you want to do is show the `500` error page, you shouldn't need to catch it at all - it will do that anyway as long as `FLASK_DEBUG=false`.

In the [www.ubuntu.com implementation](https://github.com/canonical-websites/www.ubuntu.com/blob/master/webapp/views.py#L78) I do catch `Exception` and then do `traceback.print_exc()` manually, so that I can pass the error message through to the template. I did this so that I could pass the error message through to the `500` template, just so it would be super obvious if the API key is missing (an easy mistake to make).

But a better approach would probably be to raise and catch a custom error for the case where the API key is missing, so I'm not catching `Exception`. I'll try to submit a PR to this branch to illustrate what I mean.

Reply:
Yeah, we could probably do with registering exception handlers instead. I swapped it quickly without thinking much, as Juan said we would rather have a more generic error message. I can update it as part of this work if you want.

Reply:
I've coed this up here: https://github.com/jpmartinspt/docs.snapcraft.io/pull/1

Reply:
Now using custom exception, but error is being handled by the flask error handler.
Enabled debugged application in development mode.